rep-0151.rst

```diff
@@ -0,0 +1,261 @@
+Rosdep currently calls the command `pip`.
+It might use Python2 when using the debian package Python-pip, but it could use
+Python3 inside a venv.
+
+`rosdep` could default to `sys.executable -m pip` to use the same version of
+Python as rosdep, but a user building the ROS 1 bridge for Melodic is likely
+using the debian package `Python3-rosdep`.
+To support this, rosdep should use `Python$ROS_PYTHON_VERSION -m pip`.
+Since `rosdep` sets `ROS_PYTHON_VERSION` for itself if unset, this will default
+to the same version of Python as rosdep if `ROS_PYTHON_VERSION` is unset, but
+it can still be overridden automatically when a ROS workspace is sourced.
+
+ROS packages support for Python 3
+---------------------------------
+
+There are many ROS packages using Python that will need to be modified to
+support Python 3.
+Packages using different branches for different ROS distros can drop support
+for Python 2 in their Noetic branch.
+Packages which use the same branch in multiple ROS distros may need to support
+both Python 2 and Python 3 at the same time.
+This section describes what needs to be done in both cases.
+
+Shebangs and reliance on the Python command
+'''''''''''''''''''''''''''''''''''''''''''
+Python scripts on unix systemx typically have shebang lines written as:
+
+.. code-block: bash
+
+    #!/usr/bin/env Python
+
+PEP 394 recommends distributed Python scripts to use either `Python2` or 
+`Python3` [7]_.
+The `Python` command cannot be trusted to a spefic Python version.
+On older ROS distros, scripts can continue to use `Python` since they're known
+to work on those platforms.
+In preparation for Noetic, these shebang's should be rewritten to the specific
+version of Python supported, `Python3`.
+
+Packages using the same branch will need to conditionally rewrite the shebangs.
+Packages can use the CMake macro `catkin_install_Python()` to install Python
+scripts with rewritten shebangs.
+
+The same issue appears in scripts that call the `Python` command directly.
+If they are Python scripts, they should be modified to invoke `sys.executable`.
+Otherwise, they should be edited to invoke the specific version of Python they
+require, or `Python$ROS_PYTHON_VERSION` if the script works with both.
+
+Dependencies and package.xml
+''''''''''''''''''''''''''''
+
+On platforms where the target version of Python is 2, the package.xml of a ROS
+package must refer to Python 2 dependencies, and when the target Python
+version is 3 it must refer to Python 3 dependencies.
+Packages which release from different branches for each ROS distro can replace
+rosdep keys that resolve to Python 2 dependencies to ones that resolve to
+Python 3 equivalents.
+Packages using the same code base for multiple ROS distros should instead use
+conditional dependencies as described in REP 149 [5]_.
+
+.. code-block: xml
+
+    <depend condition="$ROS_PYTHON_VERSION == '2'">Python-numpy</depend>
+    <depend condition="$ROS_PYTHON_VERSION == '3'">Python3-numpy</depend>
```

Comment:
The rosdep keys are lower case.

Reply:
Oops, I was a bit careless in using `%s/python/Python/g`. Fixed in 1c200a7